MPES: new concepts from NIAC discussions, searchable fields

lukaspie opened the pull request:
- Explicitly name the following axes in `NXmpes/NXentry/data(NXdata)`:
  - energy
  - photon energy
  - kx, ky, kz
  - k_parallel, k_perpendicular
  - angular0, angular1
  - spatial0,spatial1
  - delay
  - temperature

- Explicitly name the following axes in `NXmpes/NXinstrument/NXelectronanalyser/NXdetector/raw_data(NXdata)`:
  - pixel_x, pixel_y
  - photon energy
  - kx, ky, kz
  - k_parallel, k_perpendicular
  - angular0, angular1
  - spatial0,spatial1
  - delay
  - temperature

- NXprocess and subclasses are stored directly under `Nxmpes/NXentry`
  - `AXIS_axis_calibration` for general calibration on all axes that are defined in `NXdata`
  - Brought in special calibrations from existing base class `NXprocess_mpes` 

- As a consequence, remove base classes `NXdata_mpes` ,  `NXdata_mpes_detector`, and `NXprocess_mpes` 

- Allow for multiple beams and sources
  - `beam_probe`/`source_probe`
  - `beam_pump`/`source_probe`
  - `beamTYPE`/`sourceTYPE` -> I still think this is valuable, see the description in its docs

- change all instances of `NXidentifier` to fields inheriting from `identifierNAME`

- remove `NXsubstance`, use existing concepts in `NXsample` for now
  - use `NXsample/chemical_formula`

- add field `core_levels` under ENTRY 
  - list like `NXdata/@axes`
  - suggested enumeration of all possible core levels

- make all `NXtransformations/@vector` attributes `NX_NUMBER`

- add explicit open enumeration for `function_type` for the peak and background in `NXxps/NXfit`

Open questions:
- Do we need `k_parallel`, `k_perpendicular`?
- Is `AXIS_axis_calibration` good enough or do we need individual named calibrations for each axis?
- Can we improve core_level description?

rettigl commented on contributed_definitions/NXmpes_arpes.nxdl.xml:
```diff
@@ -357,59 +355,59 @@ with higher granularity in the data description.
             <attribute name="angular1_indices"/>
```
This should be type=NX_INT

rettigl replied:
Also above

lukaspie replied:
Thanks, all implemented

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -861,58 +1437,180 @@
                     </doc>
                 </attribute>
             </field>
             <attribute name="energy_indices" recommended="true"/>
```
This should be type=NX_INT

rettigl reviewed (commented):
_indices are INT

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -416,15 +772,177 @@
+                                This is typically the slower scan axis compared to ``spatial0``.
+                            </doc>
+                        </field>
+                        <field name="delay" type="NX_NUMBER" units="NX_TIME">
```
make optional

lukaspie replied:
done

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -416,15 +772,177 @@
+                                Calibrated delay time.
+                            </doc>
+                        </field>
+                        <field name="temperature" type="NX_NUMBER" units="NX_TIME">
```
make optional

lukaspie replied:
done

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -52,76 +52,332 @@
+                .. _Clause 11: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:sec:11
+            </doc>
+        </field>
+        <field name="core_levels">
```
make optional

lukaspie replied:
done

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -52,76 +52,332 @@
+                The convention for writing the core levels is to write the element symbol, separated by a whitespace from the electronic orbital configuration of the electronic level.
+                For example, if the C 1s and Ru 3d are probed (which have a strong overlap), one should write core_levels=["C 1s", "Ru 3d"].
+                
+                The core levels should be selected from the following list.
+                       
+                - H 1s
```
I would remove the list, and just add a few examples. Or make it an enum...

lukaspie replied:
I am also wondering if we should remove `NXelectron_level` alltogether then. We only use this in the `energy_referecing(NXcalibration)` class. We could just replace this by an `NX_char` using the same what we have in the `t ransitions` field here.

It is a bit of a pity because `NXelectron_level` had enumerations that could be checked against. Now we just have a list of strings. But I don't really see how we can make all possible notations consistent in a class. Even if we have a class, how  would one include these in the appdefs? I don't think we want to have multiple instances of  `NXelectron_level` directly under `NXentry`.

rettigl replied:
You also mentioned that this is debated in NIAC. So I also find it okay, to remove the class. Keep it simple. With explicit docstrings this is well defined enough, I think.

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -861,58 +1437,180 @@
+                    This is typically the slower scan axis compared to ``spatial0``.
+                </doc>
+            </field>
+            <field name="delay" type="NX_NUMBER" units="NX_TIME">
```
make optional

lukaspie replied:
done

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -861,58 +1437,180 @@
+                    Calibrated delay time.
+                </doc>
+            </field>
+            <field name="temperature" type="NX_NUMBER" units="NX_TIME">
```
make optional

lukaspie replied:
done

rettigl commented:
> NXprocess and subclasses are stored directly under `Nxmpes/NXentry`
> 
>     * `AXIS_axis_calibration` for general calibration on all axes that are defined in `NXdata`
> 
>     * Brought in special calibrations from existing base class `NXprocess_mpes`

Base class inheritance does not work yet in pynxtools, so I cannot test. But seems fine to me.

rettigl commented:
> * change all instances of `NXidentifier` to fields inheriting from `identifierNAME`

Is this supported already by pynxtools?

rettigl commented:
> * make all `NXtransformations/@vector` attributes `NX_NUMBER`

I was wondering if this is actually explicitly necessary, or if this is just a bug in pynxtools. In principle, the type should be inherited from the base class, no?

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -481,88 +818,141 @@ NXmpes(NXobject):
+    (NXcalibration):
+      exists: optional
+      doc: |
+        Any further calibration procedures, e.g. axis calibrations.
```
Axis calibrations are already covered by AXIS_axis_calibration. What would be another example?

lukaspie replied:
Shall we just say "Any further calibration procedures". I guess you could also calibrate the intensity (e.g., by dividing by some base line).

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -808,26 +1174,147 @@ NXmpes(NXobject):
+        exists: optional
+        unit: NX_TIME
+        doc: |
+          Calibrated delay time.
```
```suggestion
          Calibrated pump-probe delay time. Could be a link to /entry/instrument/beam_pump/pulse_delay.
```

lukaspie replied:
implemented

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -808,26 +1174,147 @@ NXmpes(NXobject):
+        doc: |
+          Calibrated photon energy of the incoming probe beam.
+          
+          Could be a link to /entry/instrumen/beam_probe/incident_energy.
```
```suggestion
          Could be a link to /entry/instrument/beam_probe/incident_energy.
```

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -21,6 +21,8 @@ NXxps(NXmpes):
         .. _ISO 18115-1:2023: https://www.iso.org/standard/74811.html
         .. _Clause 11: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:sec:11
+    core_levels:
+      exists: recommended
```
transitions

lukaspie replied:
Changed, thanks for spotting

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -280,35 +237,166 @@ NXxps(NXmpes):
+                  This function is especially valuable when standard background approximations like **Shirley or Linear backgrounds** constrain
+                  the data to a specific range, making infinite asymmetric tails impractical.
+                  
+                  The functional form depends on the implementation and should be given by the ``formula_description`` field, if possible.
```
I did not check this part.

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -342,21 +431,103 @@ NXxps(NXmpes):
+                  - :math:`A` is the amplitude,
+                  - :math:`x_0` is the binding energy where the step up begins,
+                  - :math:`\sigma` is the width of the transition,
+                  - :math:`\text{erf}(x)` is the error function.
```
I also did not check this.

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -371,17 +542,65 @@ NXxps(NXmpes):
+            enumeration: [[0, 0, -1]]
+          \@depends_on:
+            enumeration: [sample_normal_polar_angle_of_tilt]
+        sample_normal_polar_angle_of_tilt(NX_NUMBER):
```
I find the notion of "tilt" in any of these angles confusing.

lukaspie replied:
This is just directly taken from the XPS ISO standard, so I would rather not change this.

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -371,17 +542,65 @@ NXxps(NXmpes):
+          relative to a defined coordinate system.
+      transformations(NXtransformations):
+        exists: recommended
+        sample_rotation_angle(NX_NUMBER):
```
This would be azimuth for for me...

rettigl commented on contributed_definitions/nyaml/NXxps.yaml:
```diff
@@ -371,17 +542,65 @@ NXxps(NXmpes):
+        doc: |
+          Reference to the transformation describing the orientation of the sample
+          relative to a defined coordinate system.
+      transformations(NXtransformations):
+        exists: recommended
+        sample_rotation_angle(NX_NUMBER):
+          unit: NX_ANGLE
+          doc: |
+            Clockwise rotation about the sample normal.
+          \@transformation_type:
+            enumeration: [rotation]
+          \@vector(NX_NUMBER):
+            enumeration: [[0, 0, -1]]
+          \@depends_on:
+            enumeration: [sample_normal_polar_angle_of_tilt]
+        sample_normal_polar_angle_of_tilt(NX_NUMBER):
+          unit: NX_ANGLE
+          doc: |
+            Polar tilt of the sample with respect to the upward z-direction, defined by
+            the sample stage.
+          \@transformation_type:
+            enumeration: [rotation]
+          \@vector(NX_NUMBER):
+            enumeration: [[0, -1, 0]]
+          \@depends_on:
+            enumeration: [sample_normal_tilt_azimuth_angle]
+        sample_normal_tilt_azimuth_angle(NX_NUMBER):
+          unit: NX_ANGLE
+          doc: |
+            Azimuthal rotation of the sample from the y-direction towards the operator,
+            defined by the sample stage.
+          \@transformation_type:
+            enumeration: [rotation]
+          \@vector(NX_NUMBER):
+            enumeration: [[0, 0, -1]]
+          \@depends_on:
+            doc: |
+              This should point to the last element of the coordinate system transformations defined in
+              /entry/geometries/xps_coordinate_system/coordinate_system_transformations.
```
This still reads rather confusing for me, but if you are sure this corresponds to your definitions, fine. I remember we discussed about this, and you just use different names and conventions...

lukaspie replied:
This is just directly taken from the XPS ISO standard, so I would rather not change this. It is also displayed properly using the `nexus3d` tool.

lukaspie commented on contributed_definitions/nyaml/NXfit_function.yaml:
```diff
@@ -8,19 +8,43 @@ doc: |
-  (NXfit_parameter):
+      # This should be a python parsable function. Here we should provide which keywords are available
+      # and a BNF of valid grammar.
+  fit_parameters(NXparameters):
```
Yes, see https://github.com/nexusformat/definitions/blob/main/base_classes/NXparameters.nxdl.xml

lukaspie commented on contributed_definitions/nyaml/NXfit_function.yaml:
```diff
@@ -8,19 +8,43 @@ doc: |
-      This should be a python parsable function. Here we should provide which keywords are available
-      and a BNF of valid grammar.
-  (NXfit_parameter):
+      # This should be a python parsable function. Here we should provide which keywords are available
+      # and a BNF of valid grammar.
```
The thing is, NIAC does not want _any_ math in their files (hence the renaming to `formula_description`), at least for now. So, even if we provide a proper grammar, this will probably not be allowed to go into the file because it suggests that you can have "safe" math notation. So until NIAC makes a decision whether they want to include a more rigorous notation, my suggestion is to leave it just as a comment.

rettigl commented:
This I don't understand:

```
DEBUG: ===== GROUP (//entry/instrument/beam_pump [NXmpes_arpes::/NXentry/NXinstrument/beam_pump]): <HDF5 group "/entry/instrument/beam_pump" (11 members)>
DEBUG: classpath: ['NXentry', 'NXinstrument']
DEBUG: NOT IN SCHEMA
```

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -152,15 +276,17 @@ NXmpes(NXobject):
+          Properties of the probe beam at a given location.
+          
+          This is the beam that is used to facilitate the photoemission during MPES
+          experiments. It is used to measure the state of a system at a specific moment
+          in time.
```
I agree, just removed it from here.

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -184,19 +310,95 @@ NXmpes(NXobject):
             
             Example:
               * /entry/instrument/beam_probe = /entry/instrument/source_probe
+      beam_pump(NXbeam):
```
Anything else from https://github.com/nexusformat/definitions/pull/1408/files? This is the accepted version of `NXbeam` by the way

rettigl replied:
I think this is fine.

rettigl commented:
> This I don't understand:
> 
> ```
> DEBUG: ===== GROUP (//entry/instrument/beam_pump [NXmpes_arpes::/NXentry/NXinstrument/beam_pump]): <HDF5 group "/entry/instrument/beam_pump" (11 members)>
> DEBUG: classpath: ['NXentry', 'NXinstrument']
> DEBUG: NOT IN SCHEMA
> ```
> 
> ```
> DEBUG: ===== GROUP (//entry/instrument/source_pump [NXmpes_arpes::/NXentry/NXinstrument/source_pump]): <HDF5 group "/entry/instrument/source_pump" (6 members)>
> DEBUG: classpath: ['NXentry', 'NXinstrument']
> DEBUG: NOT IN SCHEMA
> ```

I had them written still as beamTYPE[beam_pump], however that did not give any warnings. We should add a check for the case that a variadic concept name in the template is invalid. I will make an issue.

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -184,19 +310,95 @@ NXmpes(NXobject):
+            This should be specified if an associated source exists.
+            
+            Example:
+              * /entry/instrument/beam_pump = /entry/instrument/source_pump
```
see above

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -184,19 +310,95 @@ NXmpes(NXobject):
+        nameType: partial
+        exists: optional
+        doc: |
+          Properties of any other photon beam at a given location.
```
Changed to `Properties of any other beam at a given location.`

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -184,19 +310,95 @@ NXmpes(NXobject):
+            This should be specified if an associated source exists.
+            
+            Example:
+              * /entry/instrument/beam_laser = /entry/instrument/source_laser
```
see above

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -321,39 +523,174 @@ NXmpes(NXobject):
             raw(NX_NUMBER):
               doc: |
                 Raw data before calibration.
+            pixel_x(NX_FLOAT):
+              exists: optional
+              unit: NX_ANY
+              doc: |
+                Detector pixel in x direction.
+            pixel_y(NX_FLOAT):
+              exists: optional
+              unit: NX_ANY
+              doc: |
+                Detector pixel in y direction.
```
changed

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -481,88 +818,141 @@ NXmpes(NXobject):
+        doc: |
+          Reference peak that was used for the calibration.
+          
+          For example: adventitious carbon | C-C | metallic Au | elemental Si | Fermi edge | vacuum level
```
You mean `level` and `reference_peak`? One is the electronic level (e.g. C 1s), the other is the peak of the actual chemical species you use in that region (e.g., the C-C peak in the C 1s region).

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -481,88 +818,141 @@ NXmpes(NXobject):
+        \@signal:
+          enumeration: [relative_intensity]
+        \@axes:
+          enumeration: [kinetic_energy]
```
This is really dependant on the kinetic energy of the electrons, this is why I went with `kinetic_energy`. Binding energy doesn't really make sense here. It is also plotted as KE vs. relative_intensity.

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -481,88 +818,141 @@ NXmpes(NXobject):
+      exists: optional
+      doc: |
+        Any further calibration procedures, e.g. axis calibrations.
+    (NXfit):
```
Yes

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -184,19 +310,95 @@ NXmpes(NXobject):
             This should be specified if an associated source exists.
             
             Example:
               * /entry/instrument/beam_probe = /entry/instrument/source_probe
```
see above, used `* /entry/instrument/beam_probe/associated_source = /entry/instrument/source_probe`

lukaspie commented on contributed_definitions/nyaml/NXfit_function.yaml:
```diff
@@ -8,19 +8,43 @@ doc: |
+      
+      Examples include "Gaussian" and Lorentzian.
+      
+      Application definitions may limit the allowed fit functions by using an enumeration for the ``function_type`` field.
```
I would just let the application definition define which function_types are allowed.

I added a bit of description that you should `formula_description` for complicated functions.

The search you suggested would be indeed something that would be helpful for me. Hence the enumeration in `NXxps`.

lukaspie commented:
> > Explicitly name the following axes in `NXmpes/NXinstrument/NXelectronanalyser/NXdetector/raw_data(NXdata)`:
> > ```
> > * pixel_x, pixel_y
> > 
> > * photon energy
> > 
> > * kx, ky, kz
> > 
> > * k_parallel, k_perpendicular
> > 
> > * angular0, angular1
> > 
> > * spatial0,spatial1
> > 
> > * delay
> > 
> > * temperature
> > ```
> 
> Should we also add energy here?

`energy` was here already, I just put here the ones that I added additionally.

> > change all instances of NXidentifier to fields inheriting from identifierNAME
>
> Is this supported already by pynxtools?

Not yet, we need to make a number of changes when we want to update the definitions in pynxtools.

>> * make all `NXtransformations/@vector` attributes `NX_NUMBER`
>
> I was wondering if this is actually explicitly necessary, or if this is just a bug in pynxtools. In principle, the type should be inherited from the base class, no?

I am actually not sure. I think if you write `NXtransformations/@vector` without anything, you are resetting the data type of that inherited attribute. This should really raise a problem when validating the definitions, but that check is not that sophisticated. Therefore, it is better to repeat it always.

> Shall we add generic PARAMETER fields to NXcalibration, etc., for storing calibration or conversion parameters? I added an NXcollection for this, but having a proper place would maybe be better?

Not a bad idea, however this would require an additional PR to NIAC. We can add it in our repo for now and standardize it later.

rettigl commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -197,27 +316,28 @@
                 </group>
```
The identifier above still needs to be changed

rettigl commented:
I remember you saying NXcoordinate_system_set should be renamed or replaced?

rettigl commented:
Is there a convention regarding the use of en-US vs. en-GB in Nexus? Our definitions are somewhat inconsistent here, e.g. "NXelectronanalyser"  vs. "analyzer" at several places in the docstrings. I would suggest fixing this to one convention.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -321,39 +522,174 @@ NXmpes(NXobject):
+              doc: |
+                (Un)calibrated photon energy of the incoming probe beam.
+                
+                Could be a link to /entry/instrumen/beam_probe/incident_energy.
```
```suggestion
                Could be a link to /entry/instrument/beam_probe/incident_energy.
```

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -481,88 +817,149 @@ NXmpes(NXobject):
+        For energy referencing, the measured energies are corrected for the charging potential
+        (i.e., the electrical potential of the surface region of an insulating sample, caused by
+        irradiation) such that those energies correspond to a sample with no surface charge.
+        Usually, the energy axis is adjusted by shifting all energies uniformally until one
```
```suggestion
        Usually, the energy axis is adjusted by shifting all energies uniformly until one
```

rettigl commented on contributed_definitions/nyaml/NXenergydispersion.yaml:
```diff
@@ -183,187 +184,187 @@ NXenergydispersion(NXobject):
-#                      
-#                      Synonyms: FE mode
+#                     In the fixed energy (FE) mode, the intensity for one single kinetic energy is measured for a
+#                     specified time. This mode is particulary useful during setup or alignment of the
```
```suggestion
#                     specified time. This mode is particularly useful during setup or alignment of the
```

rettigl commented on contributed_definitions/nyaml/NXenergydispersion.yaml:
```diff
@@ -183,187 +184,187 @@ NXenergydispersion(NXobject):
 #         <doc>
-#              Specifies the position of the energy dispesive elemeent by pointing to the last
-#              transformation in the transformation chain in the NXtransformations group.
+#             Specifies the position of the energy dispesive elemeent by pointing to the last
```
```suggestion
#             Specifies the position of the energy dispersive element by pointing to the last
```

rettigl commented on contributed_definitions/nyaml/NXfit_function.yaml:
L7: approximative -> approximate

rettigl commented:
There are also still NXidentifiers in the base classes, e.g. NXcalibration

lukaspie commented:
> There are also still NXidentifiers in the base classes, e.g. NXcalibration

This will come with #332 (merging back from NIAC main and removing classes that were not accepted). I will rebase this branch here afterwards.

lukaspie commented:
@rettigl do we still need this comment?
![image](https://github.com/user-attachments/assets/32052f9e-928e-4a57-9300-18a2220c16ba)

rettigl commented:
> @rettigl do we still need this comment? 

Don't think so.

lukaspie commented:
> I remember you saying NXcoordinate_system_set should be renamed or replaced?

Correct, we will just not use it anymore (similar to having no grouping for `Nxprocess`). The coordinate systems are now directly under NXentry.

lukaspie commented:
> Is there a convention regarding the use of en-US vs. en-GB in Nexus? Our definitions are somewhat inconsistent here, e.g. "NXelectronanalyser" vs. "analyzer" at several places in the docstrings. I would suggest fixing this to one convention. Another example is "oxidising" vs. oxidizing.

There is not really an official guideline. Most of the documentation uses American English, but there are some other cases as well. Maybe @sanbrock has an idea?

rettigl commented:
> If you agree, I would change all of our definiitions to be en-US, as well.

Yes, please.

lukaspie commented:
> > If you agree, I would change all of our definiitions to be en-US, as well.
> 
> Yes, please.

Done, see the last two commits. Main difference was `analyser -> analyzer`. Our tools will have to adjust to that as well.

As an addition, I also removed language like "NXenergydispersion is a sublass of NXelectronanalyzer", since it is not actually a subclass (in the inheritance sense), it is just used within.

If you don't have anything else to change @rettigl, I would merge here and bring the NIAC PR uptodate as well.

lukaspie commented on contributed_definitions/NXfit.nxdl.xml:
```diff
@@ -50,26 +50,54 @@
-            <dimensions rank="dimRank"/>
+            <dimensions rank="dimRank">
+                <doc>
+                    The ``input_dependent`` field must have the same rank (``dimRank``)
```
@rettigl this was apparently lost at some point, but I readded it here

rettigl commented on contributed_definitions/nyaml/NXenergydispersion.yaml:
```diff
@@ -130,34 +130,9 @@ NXenergydispersion(NXcomponent):
     doc: |
       Individual lenses in the energy dispersive section
   (NXfabrication):
-  depends_on(NX_CHAR):
```
This comes with NXcomponent, I guess?

lukaspie replied:
yes

rettigl commented on contributed_definitions/nyaml/NXfit_function.yaml:
```diff
@@ -4,23 +4,50 @@ doc: |
+      \@description(NX_CHAR):
+        doc: |
+          A description of what this parameter represents.
+      \@min_value(NX_NUMBER):
+        doc: |
+          The minimal value of the parameter, to be used as a constraint during fitting.
+      \@max_value(NX_NUMBER):
+        doc: |
+          The maximal value of the parameter, to be used as a constraint during fitting.
```
Is there also an attribute to describe that a parameter is being held constant?

lukaspie replied:
added a `fixed(NX_BOOLEAN)` attribute

lukaspie replied:
Added a `fixed(NX_BOOLEAN)` attribute 👍

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -117,26 +182,86 @@ NXmpes(NXobject):
+          the energy of the source is not the optical power of the beam, but the energy of the electron beam
+          in a synchrotron or similar.
         type:
           enumeration: [Synchrotron X-ray Source, Rotating Anode X-ray, Fixed Tube X-ray, UV Laser, Free-Electron Laser, Optical Laser, UV Plasma Source, Metal Jet X-ray, HHG laser, UV lamp, Monochromatized electron source, other]
```
Should we just make this an open enum, and remove other?

lukaspie replied:
Ah yes, that is changed in `NXsource`, I will change it here as well.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -323,39 +524,174 @@ NXmpes(NXobject):
+                It is also possible to define ``k_parallel`` and ``k_perpendicular`` for the parallel
+                and perpendicular momenta, respectively.
+                
+                Units are typically 1/Angström.
```
I think it should be `angstrom`, lower case and without umlaut. At least pint complains about "Angström", and also "Angstrom"
```suggestion
                Units are typically 1/angstrom.
```

lukaspie replied:
Okay, makes sense. Implemented.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -323,39 +524,174 @@ NXmpes(NXobject):
+                electron's momentum along the surface inside the material is directly related to
+                its measured momentum outside the material.
+                
+                Units are typically 1/Angström.
```
see above

lukaspie replied:
implemented

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -323,39 +524,174 @@ NXmpes(NXobject):
+                which accounts for the energy shift due to the material's work function and electronic
+                structure.
+                
+                Units are typically 1/Angström.
```
see above or remove

rettigl reviewed (approved):
I now checked only the yamls and only the changes shown here, but they appear fine to me apart from a few comments.
I would suggest to implement necessary changes and update the examples meanwhile NIAC reviews this, so we can potentially identify problems by our examples.

rettigl commented:
There is still a typo "presense" in NXbeam.yaml, line 17. We could propose to fix it in this PR or elsewhere.

rettigl commented:
Similar US-EN "serlialization" instead of "serialisation" in NXcalibration.yaml, line 29

rettigl commented:
Suggestion for further changes in NXcalibration:
I suggest renaming L74: coefficients(NXparameter) to calibration_parameters(NXparameters)
This contains variadic TERM, which can be used for any kind of parameter entering the calibration. I suggest to remove the notion of polynomial function, and just state that term names can be used in the fit function. scaling_factor and offset I would put as special cases also into this container. What do you think?

lukaspie commented:
> Suggestion for further changes in NXcalibration: I suggest renaming L74: coefficients(NXparameter) to calibration_parameters(NXparameters) This contains variadic TERM, which can be used for any kind of parameter entering the calibration. I suggest to remove the notion of polynomial function, and just state that term names can be used in the fit function. scaling_factor and offset I would put as special cases also into this container. What do you think?

That is tricky because `NXcalibration` is now accepted already. We would need to deprecate `coefficients` and add a new group `calibration_parameters(NXparameters)`. Maybe this could still be done as a clean-up before the new release though. I made an [issue ](https://github.com/FAIRmat-NFDI/nexus_definitions/issues/347)and will talk to Sandor about this a bit.
